share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -17,58 +17,553 @@
+        for (CoordinatorRecord record : recordList) {  // should be single record
+            if (!(record.key().message() instanceof ShareSnapshotKey) && !(record.key().message() instanceof ShareUpdateKey)) {
+                continue;
+            }
+            SharePartitionKey mapKey = null;
+            boolean shouldIncSnapshotEpoch = false;
+            if (record.key().message() instanceof ShareSnapshotKey) {
+                ShareSnapshotKey recordKey = (ShareSnapshotKey) record.key().message();
+                responseData.setResults(Collections.singletonList(WriteShareGroupStateResponse.toResponseWriteStateResult(
+                    recordKey.topicId(), Collections.singletonList(WriteShareGroupStateResponse.toResponsePartitionResult(
+                        recordKey.partition())))));
+                mapKey = SharePartitionKey.getInstance(recordKey.groupId(), recordKey.topicId(), recordKey.partition());
+                shouldIncSnapshotEpoch = true;
+            } else if (record.key().message() instanceof ShareUpdateKey) {
+                ShareUpdateKey recordKey = (ShareUpdateKey) record.key().message();
+                responseData.setResults(Collections.singletonList(WriteShareGroupStateResponse.toResponseWriteStateResult(
+                    recordKey.topicId(), Collections.singletonList(WriteShareGroupStateResponse.toResponsePartitionResult(
+                        recordKey.partition())))));
+                mapKey = SharePartitionKey.getInstance(recordKey.groupId(), recordKey.topicId(), recordKey.partition());
+            }
+
+            if (shareStateMap.containsKey(mapKey) && shouldIncSnapshotEpoch) {
+                ShareGroupOffset oldValue = shareStateMap.get(mapKey);
+                ((ShareSnapshotValue) record.value().message()).setSnapshotEpoch(oldValue.snapshotEpoch() + 1);  // increment the snapshot epoch
+            }
+            validRecords.add(record); // this will have updated snapshot epoch and on replay the value will trickle down to the map
+        }
+
+        return new CoordinatorResult<>(validRecords, responseData);
     }
 
+    /**
+     * This method finds the ShareSnapshotValue record corresponding to the requested topic partition from the
+     * in-memory state of coordinator shard, the shareStateMap.
+     * <p>
+     * This method as called by the ShareCoordinatorService will be provided with
+     * the request data which covers only key i.e. group1:topic1:partition1. The implementation
+     * below was done keeping this in mind.
+     *
+     * @param request - WriteShareGroupStateRequestData for a single key
+     * @param offset  - offset to read from the __share_group_state topic partition
+     * @return CoordinatorResult(records, response)
+     */
     public ReadShareGroupStateResponseData readState(ReadShareGroupStateRequestData request, Long offset) {
-        throw new RuntimeException("Not implemented");
+        log.debug("Read request dump - {}", request);
```

Comment:
We already have a way to log requests with log4j (`log4j.logger.kafka.request.logger`) so we don't need to log requests here. If we want to log some other details here (like "reading share state for partitions: [...]") that would be fine